thorlcr/graph/thgraphmaster.cpp

```diff
@@ -649,11 +649,40 @@ void CMasterActivity::done()
 
 void CMasterActivity::updateFileReadCostStats()
 {
+    ThorActivityKind activityKind = container.getKind();
+
+    // Returns disk access cost as well AND UPDATES numDiskReads & diskAccessCost in the file attributes
+    auto updateFileCostStats = [activityKind](bool isIndexFile, IDistributedFile *file, CThorStatsCollection &statsCollection)
+    {
+        stat_type numDiskReads = statsCollection.getStatisticSum(StNumDiskReads);
+        cost_type diskAccessCost = 0;
+
+        IPropertyTree & fileAttr = file->queryAttributes();
+        cost_type prevReadCost = 0;
+        // Check if it is a legacy file without @readCost attribute -> calculate readCost and update file attributes
+        if (!fileAttr.hasProp("@readCost") && fileAttr.hasProp("@numDiskReads"))
+            prevReadCost = calcFileAccessCost(file, 0, fileAttr.getPropInt64("@numDiskReads", 0));
+
+        file->addAttrValue("@numDiskReads", numDiskReads);
+        if (isIndexFile && (TAKkeyedjoin == activityKind || TAKindexread == activityKind))
+        {
+            stat_type numActualDiskReads = statsCollection.getStatisticSum(StNumNodeDiskFetches)
+                                            + statsCollection.getStatisticSum(StNumLeafDiskFetches)
+                                            + statsCollection.getStatisticSum(StNumBlobDiskFetches);
+            diskAccessCost = calcFileAccessCost(file, 0, numActualDiskReads);
+        }
+        else
+            diskAccessCost = calcFileAccessCost(file, 0, numDiskReads);
+        file->addAttrValue("@readCost", prevReadCost + diskAccessCost);
```

Comment:
If we do create a "@legacyReadCost", it wouldn't be exposed to the end-users in ECL watch, limiting it's usefulness. Secondly, the legacy files do not have jhtree stats so it would not be possible to produce an accurate access cost figure.  We could reset the readCost to zero but that may also be misleading.  Should we create new jira to discuss and handle this separately.  (Note: all changes related to handling legacy file costs and dynamic access cost calculation is no longer in this jira.  These changes have been moved to another jira: https://track.hpccsystems.com/browse/HPCC-30911)

Reply:
Will add comments to https://github.com/hpcc-systems/HPCC-Platform/pull/18080